# Transcriptomics Notebook

## October 10, 2024

-   we started the transcriptomics unit

-   lots of great notes on the 01_DESeq2_explore file

-   the countsTable has all the data in it that we will work with

    -   we have to round that data since DESeq2 is wimpy

-   we looked at the Column and Row data from the counts matrix

-   made a barplot

## October 15, 2024

-   we ran the 01_DESeq2_explore code to load in our data

-   chose to compare between the 2 developmental temperatures 22 and 18

    -   we are comparing 22 to 18 (important to understand what is being compared to what)

-   pulled all the results for D22 vs D18

-   we looked at the counts per sample for a specific gene as an example

-   made plots to visualize the table data, should correlate

-   made MA plots which are log fold change vs average gene expression

    -   a visual of over dispersion

-   converted the data into a data frame so that we can plot, made a new column for significance

-   

## October 17, 2024

-   03_Contrasts
-   remember that everything is a comparison, what is being compared to what
-   looked at contrasts between the developmental temps and their final temps
-   we looked at the DEGs

## October 22, 2024

-   today we focused on making the euler plots from the DEGs

-   found the unique DEGs for the treatment groups

-   We made 2 scatter plots of the merged data of the acute thermal exposure to 28C and 33C

-   the DEGs are the statistically significant points, colored to match which treatment group

## October 24, 2024

-   analyzing and visualizing gene correlation networks

-   all about WGNCA

    -   setting it up

        -   step 1 import counts matrix

        -   step 2 detect outliers

        -   step 3 normalization

        -   step 4 network construction

## October 29, 2024

-   WGCNA continued

    -   step 5 explore module eigengenes

    -   step 6 correlation of modules with traits

    -   made a heatmap of correlations

## October 31, 2024

-   working on HW #2
