# Transcriptomics Notebook

## October 10, 2024

-   we started the transcriptomics unit

-   lots of great notes on the 01_DESeq2_explore file

-   the countsTable has all the data in it that we will work with

    -   we have to round that data since DESeq2 is wimpy

-   we looked at the Column and Row data from the counts matrix

-   made a barplot

-   

## October 15, 2024

-   we ran the 01_DESeq2_explore code to load in our data

-   chose to compare between the 2 developmental temperatures 22 and 18

    -   we are comparing 22 to 18 (important to understand what is being compared to what)

-   pulled all the results for D22 vs D18

-   we looked at the counts per sample for a specific gene as an example

-   made plots to visualize the table data, should correlate

-   made MA plots which are log fold change vs average gene expression

    -   a visual of over dispersion

-   converted the data into a data frame so that we can plot, made a new column for significance

-   

## October 24, 2024

-   we worked with WGNCA

-   imported our counts data

-   loaded in our trait data

-   used goodSampleGenes to see how many of the total genes were good quality, looked like all

    -   deeper dive using gsg\$goodGenes showed that not all were good

    -   
